doc/source/advanced.rst

```diff
@@ -0,0 +1,381 @@
+
+.. code-block:: python
+
+    >>> bucket = b2_api.get_bucket_by_name(bucket_name)
+    >>> input_sources = [
+    ...     RemoteFileUploadSource('4_z5485a1682662eb3e60980d10_f113f963288e711a6_d20190404_m065910_c002_v0001095_t0044', offset=100, length=100),
+    ...     LocalUploadSource('my_local_path/to_file.txt'),
+    ...     RemoteFileUploadSource('4_z5485a1682662eb3e60980d10_f1022e2320daf707f_d20190620_m122848_c002_v0001123_t0020', length=2123456789),
+    ... ]
+    >>> file_info = {'how': 'good-file'}
+    >>> bucket.concatenate_stream(input_sources, remote_name, file_info)
+    <b2sdk.file_version.FileVersionInfo at 0x7fc8cd560551>
+
+
+
+Concatenate files of unknown size
+=================================
+
+While it is supported by B2 server, this pattern is currently not supported by **b2sdk**. If you think there is a good reason why **b2sdk** should support it, please `open a github issue <https://github.com/Backblaze/b2-sdk-python/issues/new>`_ and describe your use case.
+
+
+*********************
+Synthethize an object
+*********************
+
+Using methods described below an object can be created from both local and remote sources while avoiding downloading small ranges (caused by :term:`absoluteMinimumPartSize` limitation) when such range is already present on a local drive.
+
+
+Update a file efficiently
+====================================
+
+:meth:`b2sdk.v1.Bucket.create_file` accepts an iterable which *can contain overlapping destination ranges*.
+
+.. note::
+  Following examples *create* new files - data in bucket is immutable, but **b2sdk** can create a new file version with the same name and updated content
+
+
+Append to the end of a file
+---------------------------
+
+The assumption here is that the local file has been appended to since it was last uploaded. This assumption is verified by **b2sdk** when possible by recalculating checksums of the overlapping remote and local ranges. If copied remote part checksum does not match the locally available source, file creation process is interrupted and an exception is raised.
+
+.. code-block:: python
+
+    >>> bucket = b2_api.get_bucket_by_name(bucket_name)
+    >>> input_sources = [
+    ...     WriteIntent(
+    ...         data=RemoteFileUploadSource(
+    ...             '4_z5485a1682662eb3e60980d10_f113f963288e711a6_d20190404_m065910_c002_v0001095_t0044',
+    ...             offset=0,
+    ...             length=5000000,
+    ...         ),
+    ...         destination_offset=0,
+    ...     ),
+    ...     WriteIntent(
+    ...         data=LocalFileUploadSource('my_local_path/to_file.txt'),  # of length 60000000
+    ...         destination_offset=0,
```

Comment:
Should the offset be `5000000`?

Reply:
In this example the ranges of the remote file (0-5000000) and local file (0-60000000) are partially overlapping.

If offset of the local source would be 5000000, it would mean that the caller would like to concatenate the remote file with a local file and it would be possible to do it using `create_file`, but in such case `concatenate` could be used too and it has a more pleasant interface (where one doesn't need to worry about manually providing offsets).